webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts

```diff
@@ -173,65 +173,81 @@ function myAudits(state = [], action: GenericAction) {
     }
 }
 
+function receiveUserProfile(state: IDMappedObjects<UserProfile>, received: UserProfile) {
+    const existing = state[received.id];
+
+    if (!existing) {
+        // No existing data to merge with
+        return {
+            ...state,
+            [received.id]: received,
+        };
+    }
+
+    const merged = {
+        ...existing,
+        ...received,
+    };
+
+    // MM-53377:
+    // For non-admin users, certain API responses don't return details for the current user that would be sanitized
+    // out for others. This currently includes:
+    // - email (if PrivacySettings.ShowEmailAddress is false)
+    // - first_name/last_name (if PrivacySettings.ShowFullName is false)
+    // - last_password_update
+    // - auth_service
+    // - notify_props
+    //
+    // Because email, first_name, last_name, and auth_service can all be empty strings regularly, we can't just
+    // merge the received user and the existing one together like we normally would. Instead, we can use the
+    // existence of existing.notify_props or existing.last_password_update to determine which object has that extra
+    // data so that it can take precedence. Those fields are:
+    // 1. Never empty or zero by Go standards
+    // 2. Only ever sent to the current user, not even to admins, so we know that the object contains privileged data
+    //
+    // Note that admins may have the email/name/auth_service of other users loaded as well. This does not prevent that
+    // data from being replaced when merging sanitized user objects. There doesn't seem to be a way for us to detect
+    // whether the object is sanitized for admins.
+    if (existing.notify_props && (!received.notify_props || Object.keys(received.notify_props).length === 0)) {
+        merged.email = existing.email;
+        merged.first_name = existing.first_name;
+        merged.last_name = existing.last_name;
```

Comment:
If `PrivacySettings.ShowEmailAddress` or `PrivacySettings.ShowFullName` are `true`, are we now discarding legitimate changes to this information?

Reply:
No, that shouldn't be possible. With the if statement above, the only case where that would be possible would be if the response or websocket event for the current user changing their name/email was sanitized. From what I can tell, that does not happen because `app.UpdateUser` uses the same mostly-unsanitized User object for both the response to the API request and for the WebSocket message that's broadcast to the affected user